### PR TITLE
feat(query): support array_sort

### LINKED ISSUE #9940: array_sort(array)

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

| array_sort(array, ('ASC'\|'DESC'), ('NULLS FIRST'\|'NULLS LAST'))	| Sorts the elements of the array. |	array_sort([3, 6, 1, 2]) |	[1, 2, 3, 6] |
-- | -- | -- | --

Closes #9940

### CODE REVIEW

File: src/query/functions/src/scalars/array.rs [hunk 1]
```diff
@@ -134,6 +134,26 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+    );
+
+    registry.register_passthrough_nullable_1_arg::<ArrayType<GenericType<0>>, ArrayType<GenericType<0>>, _, _>(
+        "array_sort",
+        FunctionProperty::default(),
+        |_| FunctionDomain::Full,
+        vectorize_1_arg::<ArrayType<GenericType<0>>, ArrayType<GenericType<0>>>(|arr, _| {
+            let data_type = arr.data_type();
+            let mut builder = ColumnBuilder::with_capacity(&data_type, arr.len());
+            arr.iter().sorted().for_each(|val| { builder.push(val) });
```
Comment (b41sh) [inline src/query/functions/src/scalars/array.rs:151]: better to use the `DataBlock::sort` function.

File: src/query/functions/src/scalars/array.rs [hunk 2]
```diff
@@ -134,6 +137,76 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+        |_| FunctionDomain::Full,
+        vectorize_1_arg::<EmptyArrayType, EmptyArrayType>(|arr, _| arr),
+    );
+
+    registry.register_passthrough_nullable_1_arg::<ArrayType<GenericType<0>>, ArrayType<GenericType<0>>, _, _>(
+        "array_sort",
+        FunctionProperty::default(),
+        |_| FunctionDomain::Full,
+        vectorize_1_arg::<ArrayType<GenericType<0>>, ArrayType<GenericType<0>>>(|arr, _| {
+            let data_type = arr.data_type();
```
Comment (b41sh) [inline src/query/functions/src/scalars/array.rs:152]: we can check the length of `arr`, if the number of elements is zero or one, don't need to be sorted.

**Comment by TCeason** [comment]
@soyeric128 Please add this func into array doc.

File: src/query/ast/tests/it/testdata/expr.txt [hunk 3]
```diff
@@ -1646,6 +1646,109 @@ Substring {
...
+    },
+    asc: true,
+    null_first: true,
+}
+
+
+---------- Input ----------
+array_sort([2,0.1], 'ASC')
+---------- Output ---------
+ARRAY_SORT([2, 0.1]) , ASC , NULLS FIRST)
```
Comment (andylokandy) [inline src/query/ast/tests/it/testdata/expr.txt:1681]: Display is not quoted

File: src/query/functions/src/scalars/array.rs [hunk 4]
```diff
@@ -744,4 +754,39 @@ fn register_array_aggr(registry: &mut FunctionRegistry) {
             }))
         });
     }
+
+    for (fn_name, sort_desc) in ARRAY_SORT_FUNCTIONS {
```
Comment (andylokandy) [inline src/query/functions/src/scalars/array.rs:758]: Why not to
```suggestion
    for (fn_name, sort_desc) in [
    ("array_sort_asc_null_first", (true, true)),
    ("array_sort_desc_null_first", (false, true)),
    ("array_sort_asc_null_last", (true, false)),
    ("array_sort_desc_null_last", (false, false)),
] {
```

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 5]
```diff
@@ -1738,6 +1749,28 @@ impl<'a> TypeChecker<'a> {
...
+        let box (arg, _type) = self.resolve(expr, None).await?;
+        let func_name = if *asc && *null_first {
+            "array_sort_asc_null_first"
+        } else if *asc && !(*null_first) {
+            "array_sort_asc_null_last"
+        } else if !(*asc) && *null_first {
+            "array_sort_desc_null_first"
+        } else {
+            "array_sort_desc_null_last"
+        };
```
Comment (andylokandy) [inline src/query/sql/src/planner/semantic/type_check.rs:1769]: prefer to use match

File: src/query/functions/tests/it/scalars/parser.rs [hunk 6]
```diff
@@ -279,6 +279,28 @@ pub fn transform_expr(ast: AExpr, columns: &[(&str, DataType)]) -> RawExpr {
...
+        } => {
+            let name = if asc && null_first {
+                "array_sort_asc_null_first".to_string()
+            } else if asc && !null_first {
+                "array_sort_asc_null_last".to_string()
+            } else if !asc && null_first {
+                "array_sort_desc_null_first".to_string()
+            } else {
+                "array_sort_desc_null_last".to_string()
+            };
```
Comment (andylokandy) [inline src/query/functions/tests/it/scalars/parser.rs:296]: prefer to use match

File: src/query/functions/src/scalars/array.rs [hunk 7]
```diff
@@ -744,4 +754,39 @@ fn register_array_aggr(registry: &mut FunctionRegistry) {
...
+                let len = columns.get(0).map_or(1, |c| c.len());
+                let columns = columns
+                    .iter()
+                    .map(|col| BlockEntry {
+                        data_type: col.data_type(),
+                        value: Value::Column(col.clone()),
+                    })
+                    .collect();
+                let sort_block = DataBlock::sort(&DataBlock::new(columns, len), &sort_desc, None).unwrap();
+                sort_block.columns()[0].value.convert_to_full_column(&data_type, 1)
```
Comment (andylokandy) [inline src/query/functions/src/scalars/array.rs:787]: why having to `convert_to_full_column`?

File: src/query/functions/src/scalars/array.rs [hunk 8]
```diff
@@ -744,4 +754,38 @@ fn register_array_aggr(registry: &mut FunctionRegistry) {
...
+            FunctionProperty::default(),
+            |_| FunctionDomain::Full,
+            vectorize_1_arg::<ArrayType<GenericType<0>>, ArrayType<GenericType<0>>>(|arr, _| {
+                let sort_desc = vec![SortColumnDescription {
+                    offset: 0,
+                    asc: sort_desc.0,
+                    nulls_first: sort_desc.1,
+                }];
+                let columns = vec![arr];
+                let len = columns.get(0).map_or(1, |c| c.len());
```
Comment (b41sh) [inline src/query/functions/src/scalars/array.rs:777]: Can use `let len = arr.len();`.
Also check if length is zero or one to rturn directly.

File: src/query/functions/src/scalars/array.rs [hunk 9]
```diff
@@ -744,4 +754,38 @@ fn register_array_aggr(registry: &mut FunctionRegistry) {
...
+                let columns = columns
+                    .iter()
+                    .map(|col| BlockEntry {
```
Comment (b41sh) [inline src/query/functions/src/scalars/array.rs:780]: `let columns = vec![BlockEntry{…}];`

File: src/query/ast/src/ast/format/syntax/expr.rs [hunk 10]
```diff
@@ -318,6 +318,38 @@ pub(crate) fn pretty_expr(expr: Expr) -> RcDoc<'static> {
...
+            asc,
+            null_first,
+            ..
+        } => {
+            let res = pretty_expr(*expr);
+            if asc {
+                res.clone()
+                    .append(RcDoc::text(","))
+                    .append(RcDoc::space())
+                    .append(RcDoc::text("ASC"));
```
Comment (andylokandy) [inline src/query/ast/src/ast/format/syntax/expr.rs:332]: we may also need to quote them

**Review by andylokandy — APPROVED** [review]

**Review by b41sh — APPROVED** [review]

**Pull request merged**